# Arithmetic Operations Using NumPy

## __`Addition`__

In [1]:
import numpy as np
a = np.array([30,20,10])
b = np.array([10,20,30])
result = np.add (a,b)
print(result)

[40 40 40]


## __`Subtraction`__

In [2]:
# Perform element-wise subtraction of two arrays using 'np.subtract' method
a = np.array([[30,40,60], [50,70,90]])
b = np.array([[10,20,30], [40,30,80]])
result = np.subtract (a,b)
print(result)

[[20 20 30]
 [10 40 10]]


## __`Multiply`__

In [3]:
a = np.array([30,20,10])
b = np.array([10,20,30])
result = np.multiply (a,b)
print(result)

[300 400 300]


## __`Divide`__

In [4]:
a = np.array([[1,2,1],[2,2,3]])
b = np.array([3,4,5])
result = np.divide (a,b)
print(result)

[[0.33333333 0.5        0.2       ]
 [0.66666667 0.5        0.6       ]]


## __`Power of`__

In [5]:
a = [2,2,2,2,2]
b = [2,3,4,5,6]
c = np.power(a,b)
print(c)

[ 4  8 16 32 64]


 # Statistical Function in Numpy

## __`Median`__
finds the middle value of all the numbers in the array after sorting them.

__1. Flatten the array__

```python
[4, 3, 2, 10, 1, 0, 5, 8, 24]
```

__2. Sort the values__

```python
[0, 1, 2, 3, 4, 5, 8, 10, 24]

```

__3. Pick the middle value__

4


In [7]:
array = np.array([[4,3,2],[10,1,0],[5,8,24]])
np.median(array)
print(np.median(array))

4.0


## __`Mean (μ)`__
Calculates the average value of the numbers.

__1. Takes all numbers in the array__

```python
4, 3, 2,
10, 1, 0,
5, 8, 24

```
__2. Adds them together__

```python
4 + 3 + 2 + 10 + 1 + 0 + 5 + 8 + 24 = 57

```

__3. Divides by how many numbers there are__

```python
9 numbers total

57 ÷ 9 = 6.333333333333333


```

In [8]:
np.mean(array)
print(np.mean(array))

6.333333333333333


## __`Standard Deviation σ`__
Measures how spread out the numbers are from the mean.

```python
array = np.array([[4,3,2],[10,1,0],[5,8,24]])

# Numpy treats it as one list of numbers
[4,3,2,10,1,0,5,8,24]

# there are 9 numbers

# the mean/average is 

6.333333333333333
```

__1. Finds how far each value is from the mean__

| Value | Deviation (x − μ) |
| ----- | ----------------- |
| 4     | −2.3333           |
| 3     | −3.3333           |
| 2     | −4.3333           |
| 10    | 3.6667            |
| 1     | −5.3333           |
| 0     | −6.3333           |
| 5     | −1.3333           |
| 8     | 1.6667            |
| 24    | 17.6667           |


__2. Squares those differences__

| Deviation | Squared  |
| --------- | -------- |
| −2.3333   | 5.4444   |
| −3.3333   | 11.1111  |
| −4.3333   | 18.7778  |
| 3.6667    | 13.4444  |
| −5.3333   | 28.4444  |
| −6.3333   | 40.1111  |
| −1.3333   | 1.7778   |
| 1.6667    | 2.7778   |
| 17.6667   | 312.1111 |


4. Averages them

```python

5.4444+11.1111+18.7778+13.4444+28.4444+40.1111+1.7778+2.7778+312.1111 = 434.0(approximately)

434/9 = 48.222



```
__48.22 is the <u>variance</>__

6. Takes the square root
```python


σ = √48.222 ≈ 6.944222218666553

__`σ (sigma)`__ means standard deviation

__`≈`__ (approximately equal to)
```
<u>Population vs Sample</u>

By default, NumPy calculates population standard deviation.

ddof=0 → divide by N (population)

ddof=1 → divide by N − 1 (sample)

In [9]:
# Standard Deviation
np.std(array)
print(np.std(array))

6.944222218666553


## __`Variance σ²`__
measures how spread out your data is from the mean, but before taking the square root.

| Value | Value − Mean |
| ----- | ------------ |
| 4     | −2.3333      |
| 3     | −3.3333      |
| 2     | −4.3333      |
| 10    | 3.6667       |
| 1     | −5.3333      |
| 0     | −6.3333      |
| 5     | −1.3333      |
| 8     | 1.6667       |
| 24    | 17.6667      |

| Difference | Squared  |
| ---------- | -------- |
| −2.3333    | 5.4444   |
| −3.3333    | 11.1111  |
| −4.3333    | 18.7778  |
| 3.6667     | 13.4444  |
| −5.3333    | 28.4444  |
| −6.3333    | 40.1111  |
| −1.3333    | 1.7778   |
| 1.6667     | 2.7778   |
| 17.6667    | 312.1111 |


average of the squared differences = 48.2222


In [10]:
# Variance
np.var(array)
print(np.var(array))

48.22222222222222


## __`Calculating Percentiles`__

### __`np.percentile()`__
Computes the nth percentile of the array elements(between 1-100)

It finds the value below which 50% of the data falls.
The 50th percentile is the median.

1. Flatten the array
   
```python
[4, 3, 2, 10, 1, 0, 5, 8, 24]
```
2. Sort the values

```python
[0, 1, 2, 3, 4, 5, 8, 10, 24]

```

3. Find the 50th percentile
   There are 9 values

The middle value (5th item) is: 4

In [12]:
array = np.array([[4,3,2],[10,1,0],[5,8,24]])

np.percentile(array,50)

np.float64(4.0)

## __`String Function in Numpy`__

__`char`__ IS:

A submodule inside NumPy

Accessed as __`np.char`__

__`np.char`__ is the actual name of a module inside NumPy.

    it cannot be replaced with a differend word


__`add`__ 
function that concatenates strings element-by-element

In [13]:
# Perform element-wise string concatenation for two arrays of string
x = np.array(['Hello','World'])
y = np.array(['Welcome', 'Learners'])
result = np.char.add(x,y)
print(result)

['HelloWelcome' 'WorldLearners']


__`replace`__

In [14]:
# Replacing the old substring with the new substring
str = "Hello How Are You"
print(str)
z = np.char.replace (str, 'Hello', 'Hi')
print (z)

Hello How Are You
Hi How Are You


__`upper`__

In [15]:
# Converting all lowercase characters in a string to uppercase and vice-versa
str1 = "hello how are you"
print(str1)
z = np.char.upper (str1)
print(z)
str2 = "GREETINGS OF THE DAY"
print(str2)
s = np.char.lower (str2)
print(s)

hello how are you
HELLO HOW ARE YOU
GREETINGS OF THE DAY
greetings of the day
